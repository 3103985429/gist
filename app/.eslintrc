{
  // extend root arrays
  "extends": "../.eslintrc.js",
  "parserOptions": {
    "ecmaVersion": 6,
  },
  "env": {
    "browser": false,
    "node": true
  }
}
