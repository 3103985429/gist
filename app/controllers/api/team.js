const winston = require('winston');
const async = require('async');
const { isNil, isEqual } = require('lodash');
const Batch = require('batch');
const Team = require('../../models/teams');
const User = require('../../models/users');
const s3ImageHosting = require('../../libs/utils/aws-image-hosting');
const datasource_file_service = require('../../libs/utils/aws-datasource-files-hosting');
const datasource_description = require('../../models/descriptions');
const s3FileHosting = require('../../libs/utils/aws-datasource-files-hosting');
const sample_dataset = require('./sample_dataset');
const { handleError } = require('../../utils/requests');
const { validateObjectId, validateTeamAdmin } = require('../../utils/validation');
const { RequestError, UserError } = require('../../libs/system/errors');

const ALLOWED_FILE_TYPES = /text\/plain|image\/bmp|image\/gif|image\/jpeg|image\/png|image\/svg\+xml|image\/svg xml\/application|pdf/gm;

const checkUser = next => {
    return function (req, res) {
        if (!req.user) {
            return res.status(401).send({ error: 'session expired' });
        }
        return next(req, res);
    };
};

module.exports.getAll = function (req, res) {
    Team.find({})
        .populate({
            path: 'datasourceDescriptions',
            select: { uid: 1, title: 1, replaced: 1, master_id: 1, schema_id: 1, updatedAt: 1 },
        })
        .sort({ title: 1 })
        .exec(function (err, teams) {
            if (err) {
                winston.error(err);
                res.send({ error: err.message });
            } else {
                res.status(200).json(teams);
            }
        });
};

module.exports.create = checkUser((req, res) => {
    Team.create(req.body, (err, createdTeam) => {
        if (err) {
            res.send({ error: err.message });
        } else {

            User.findById(req.user, (err, user) => {
                if (err) {
                    res.send({ error: err.message });
                } else {
                    user._team.push(createdTeam._id);
                    user.save();

                    // this will be a pain to have in dev if ever someone wants to wipe their local db, setting to
                    // production only for now also, if we're creating sampleTeam for the first time
                    if (createdTeam.title !== 'sampleTeam' && (process.env.NODE_ENV === 'production' || process.env.NODE_ENV === 'staging')) {
                        return sample_dataset.delegateDatasetDuplicationTasks(req.user, createdTeam, err => {
                            if (err) {
                                res.send({ error: err });
                            } else {
                                createdTeam.notifyNewTeamCreation();
                                res.json(createdTeam);
                            }
                        });
                    }

                    createdTeam.notifyNewTeamCreation();
                    res.json(createdTeam);
                }
            });
        }
    });
});

module.exports.search = function (req, res) {
    Team.find(req.query, (err, foundTeams) => {
        if (err) {
            return res.status(500).json({ error: err.message });
        }

        res.status(200).json(foundTeams);
    });
};

module.exports.update = checkUser((req, res) => {
    Team.findByIdAndUpdate(req.params.id)
        .exec((err, team) => {
            if (err) {
                return res.status(500).json({ error: err.message });
            }

            if (!team) {
                return res.status(404).json({ error: 'Team not found.' });
            }

            for (let attr in req.body) {
                team[attr] = req.body[attr];
            }
            team.save(err => {
                if (err) {
                    return res.status(500).json({ error: err.message });
                }

                return res.json({ team });
            });
        });
});

module.exports.signedUrlForAssetsUpload = checkUser((req, res) => {
    if (!req.query.fileType.match(ALLOWED_FILE_TYPES)) {
        return res.status(406).send({ error: 'Invalid file format' });
    }

    Team.findById(req.params.id)
        .exec((err, team) => {
            let key;
            if (req.query.assetType === 'logo' || req.query.assetType === 'logo_header') {
                key = team.subdomain + '/assets/logo/' + req.query.fileName;
            } else {
                key = team.subdomain + '/assets/' + req.query.assetType + '/' + req.query.fileName;
            }

            s3ImageHosting.signedUrlForPutObject(key, req.query.fileType, (err, data) => {
                if (err) {
                    return res.status(500).send({ error: err.message });
                } else {
                    return res.json({ putUrl: data.putSignedUrl, publicUrl: data.publicUrl });
                }
            });
        });
});

module.exports.loadIcons = checkUser((req, res) => {

    User.findById(req.user)
        .populate('defaultLoginTeam')
        .exec((err, user) => {
            if (err) {
                res.status(500).send({ error: err.message });
            } else {

                s3ImageHosting.getAllIconsForTeam(user.defaultLoginTeam.subdomain, (err, data) => {
                    if (err) {
                        res.status(500).send({ error: err.message });
                    } else {
                        res.json({ iconsUrl: data });
                    }
                });

            }
        });

});

module.exports.delete = checkUser((req, res) => {

    let batch = new Batch();
    batch.concurrency(1);

    let teamToDelete;
    let unauthorized = false;

    batch.push(done => {

        Team.findById(req.params.id, (err, team) => {
            if (err) {
                done(err);
            } else {
                teamToDelete = team;
                // if (team.admin == req.user) done(); //comment out, can team admin delete team?
                batch.push(done => {

                    User.findById(req.user, (err, user) => {
                        if (err) {
                            done(err);
                        } else {
                            if (user.isSuperAdmin() || user.superUser) {
                                done();
                            } else {
                                // todo: does it even work?
                                unauthorized = true;
                                done();
                            }
                        }
                    });

                });
                done();
            }
        });
    });

    batch.push(done => {
        if (!unauthorized) {
            datasource_description.find({ _team: teamToDelete._id }, (err, datasets) => {
                if (err) {
                    done(err);
                } else {

                    datasets.forEach(dataset => {
                        dataset.remove();
                    });

                    done();
                }
            });
        } else {
            done();
        }
    });

    batch.push(done => {
        // todo: even if user is unauthorised?
        datasource_file_service.deleteTeam(teamToDelete.subdomain, done);
    });

    batch.push(done => {
        if (!unauthorized) {
            User.find({ _team: teamToDelete._id }, (err, allusers) => {
                if (err) {
                    done(err);
                } else {

                    if (allusers) {

                        allusers.forEach(user => {
                            const index = user._team.indexOf(teamToDelete._id);
                            if (index >= 0) {
                                user._team.splice(index, 1);
                                if (user.defaultLoginTeam === teamToDelete._id.toString()) {
                                    user.defaultLoginTeam = null;
                                    if (user._team.length === 0) {
                                        user.defaultLoginTeam = null;
                                    } else {
                                        user.defaultLoginTeam = user._team[0];
                                    }
                                }
                                user.markModified('defaultLoginTeam');
                                user.markModified('_team');
                                user.save();
                            }
                        });

                        done();
                    } else {
                        done();
                    }
                }
            });
        } else {
            done();
        }
    });

    batch.push(done => {
        // todo: unauthorised?
        teamToDelete.remove(done);
    });

    batch.end(err => {
        if (err) {
            winston.error(err);
            return res.status(500).send(err);
        }
        if (unauthorized) {
            res.status(401).send({ error: 'unauthorized' });
        } else {
            return res.json({ message: 'ok' });
        }
    });
});

module.exports.deleteImage = checkUser((req, res) => {
    // only the admin or super admin can change images
    User.findById(req.user, (err, user) => {
        if (err) {
            return res.status(500).json({ error: err.message });
        }

        if (!user.isSuperAdmin() && !user._team.some(team => isEqual(team.toString(), req.params.id))) {
            return res.status(401).json({ error: 'unauthorized' });
        }

        Team.findOne({ _id: req.params.id }, (err, teamDoc) => {
            if (err) {
                return res.status(500).send(err.message);
            } else {
                const key = teamDoc.subdomain + '/assets/' + req.params.folder + '/' + req.params.filename;
                // delete the key from s3
                s3FileHosting.deleteObject(key, (err, data) => {
                    if (err) {
                        res.status(500).send({ error: err.message });
                    } else {

                        // if the url is in the database
                        if (req.params.folder === 'logo_header') {
                            teamDoc.logo_header = undefined;
                        } else if (req.params.folder === 'logo') {
                            teamDoc.logo = undefined;
                        } else {
                            return res.json({ message: 'ok' });
                        }
                        teamDoc.save();
                        return res.json({ doc: teamDoc });
                    }
                });
            }
        });
    });
});

module.exports.addAdmin = checkUser(async (req, res) => {
    const newAdminId = req.params.id;

    try {
        validateObjectId(newAdminId);

        const user = await User.findById(req.user)
            .select('_id defaultLoginTeam email superUser')
            .populate('defaultLoginTeam', 'admin')
            .lean()
            .exec();
        const team = user.defaultLoginTeam;
        const teamId = team._id.toString();

        validateTeamAdmin(user, team);

        const newAdmin = await User.findById(newAdminId)
            .select('_id _team')
            .lean()
            .exec();

        if (!newAdmin) {
            return handleError(new RequestError('User doesn\'t exist'), res);
        }

        if (!newAdmin._team.some(id => id.toString() === teamId)) {
            return handleError(new RequestError('User doesn\'t belong to the team'), res);
        }

        const admin = team.admin.concat(newAdmin._id);

        if (team.admin.some(id => id.toString() === newAdminId)) {
            return handleError(new RequestError('User is already an admin'), res);
        }

        await Team.findOneAndUpdate({ _id: teamId }, { admin }).exec();

        res.sendStatus(204);
    } catch (err) {
        handleError(err, res);
    }
});

module.exports.deleteAdmin = checkUser(async (req, res) => {
    const adminId = req.params.id;

    try {
        validateObjectId(adminId);

        const user = await User.findById(req.user)
            .select('_id defaultLoginTeam email superUser')
            .populate('defaultLoginTeam', 'admin')
            .lean()
            .exec();
        const team = user.defaultLoginTeam;

        validateTeamAdmin(user, team);

        const adminToRemove = await User.findById(adminId)
            .select('_id')
            .lean()
            .exec();

        if (!adminToRemove) {
            return handleError(new RequestError('User doesn\'t exist'), res);
        }

        if (!team.admin.some(id => id.toString() === adminId)) {
            return handleError(new RequestError('User isn\'t an admin'), res);
        }

        if (team.admin.length === 1) {
            return handleError(new UserError('Team needs to have at least one admin'), res);
        }

        const admin = team.admin.filter(id => id.toString() !== adminId);

        await Team.findOneAndUpdate({ _id: team._id }, { admin }).exec();

        res.sendStatus(204);
    } catch (err) {
        handleError(err, res);
    }
});

const getTeamBySubdomain = async ({ user: userId, subdomains: [subdomain], query: { vizName } }) => {
    let user = null;

    if ((isNil(subdomain) || subdomain === '') && !process.env.SUBDOMAIN) {
        return null;
    }

    if (userId) {
        user = await User.findById(userId).populate('_team');
    }

    return await Team.getTeamEntries(subdomain, user, vizName);
};

// team page
module.exports.getTeamBySubdomain = async.asyncify(getTeamBySubdomain);
