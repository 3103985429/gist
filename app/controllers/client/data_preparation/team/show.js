const async = require('async');
const { omit, each, orderBy, get } = require('lodash');
const queryString = require('querystring');
const ObjectId = require('mongodb').ObjectID;

const raw_source_documents = require('../../../../models/raw_source_documents');
const User = require('../../../../models/users');
const Story = require('../../../../models/stories');
const View = require('../../../../models/views');
const func = require('../../func');
const viewByHelpers = require('./viewByHelpers');
const libraries = require('../../../../libs/middleware/views/libraryList');
const { compareCreatedAt } = require('../data_prep_helpers/sort.helpers');

const INSIGHTS_LIMIT = 10;
const VISUALIZATION_LIMIT = 18;

module.exports.BindData = function (req, teamDescription, callback) {
    var team = omit(teamDescription, 'datasourceDescriptions');
    var storiesArray = [];
    var baseUrl;

    var default_customView;
    if (team.isEnterprise) {
        default_customView = team.subdomain;
    }

    var iterateeFn = async.ensureAsync(function (dataSourceDescription, cb) // prevent stack overflows from this sync iteratee
    {
        raw_source_documents.Model.findOne({
            primaryKey: dataSourceDescription._id,
        }, function (err, doc) {
            if (err) {
                return cb(err, null);
            }
            if (!doc) {
                return cb(err, null);
            }

            // construct the query
            var query = {};
            if (typeof dataSourceDescription.fe_filters.default !== 'undefined') {
                query = dataSourceDescription.fe_filters.default;
            }
            if (dataSourceDescription.importRevision > 1) {
                query.revision = dataSourceDescription.importRevision;
            }
            query = queryString.stringify(query);

            var default_view = (default_customView) ? default_customView : 'gallery';
            if (typeof dataSourceDescription.fe_views.default_view !== 'undefined' && !default_customView) {
                default_view = dataSourceDescription.fe_views.default_view;
            }

            const formatName = ({ firstName = '', lastName = '' }) => `${firstName} ${lastName}`.trim();

            var updatedByDisplayName = formatName(dataSourceDescription.updatedBy || {});
            var authorDisplayName = formatName(dataSourceDescription.author || {});

            const viewsPromise = View.getAllBuiltInViewsAsObject();

            viewsPromise.then(function(views) {
                var sourceDescription = {
                    _id: dataSourceDescription._id,
                    key: dataSourceDescription.uid,
                    sourceDoc: doc,
                    updatedAt: dataSourceDescription.updatedAt,
                    createdAt: dataSourceDescription.createdAt,
                    title: dataSourceDescription.title,
                    brandColor: dataSourceDescription.brandColor || '#005CB5',
                    description: dataSourceDescription.description,
                    urls: dataSourceDescription.urls,
                    lastUpdatedBy: updatedByDisplayName,
                    author: authorDisplayName,
                    query: query,
                    default_view: default_view,
                    banner: dataSourceDescription.banner,
                    view_visibility: func.mergeViewDisplayNamesIntoDatasourceDescription(dataSourceDescription.fe_views.views, views),
                };

                var findQuery = {
                    $and: [
                        { datasourceDescription: dataSourceDescription._id },
                        {
                            $or: [
                                { isPublic: true },
                                { isPublic: { $exists: false } },
                            ],
                        },
                    ],
                };

                if (req.user) {
                    var userId = ObjectId(req.user);
                    var userFindQuery = [
                        {
                            $and: [
                                { isPublic: false },
                                { createdBy: userId },
                            ],
                        },
                    ];

                    findQuery.$and[1].$or = findQuery.$and[1].$or.concat(userFindQuery);
                }

                Story.find(findQuery)
                    .deepPopulate('sharedPages createdBy datasourceDescription datasourceDescription._team')
                    .exec(function (err, stories) {
                        if (err) {
                            return callback(err, null);
                        }

                        if (stories && stories.length > 0) {
                            Array.prototype.push.apply(storiesArray, stories);
                            sourceDescription.storiesCount = stories.length;
                        } else {
                            sourceDescription.storiesCount = 0;
                        }

                        return cb(err, sourceDescription);
                    });
            });

        });
    });

    var completionFn = function (err, sourceDescriptions) {
        var rootDomain = process.env.HOST ? process.env.HOST : 'localhost:9080';
        baseUrl = process.env.USE_SSL === 'true' ? 'https://' : 'http://';

        if (process.env.NODE_ENV !== 'enterprise') {
            baseUrl += teamDescription.subdomain + '.';
        }

        baseUrl += rootDomain;

        // Set viewBy as an object signifying which of the tabs is to be active on page load
        let viewBy = viewByHelpers.instantiateViewBy(req.path);

        // Make sure the defaults view aren't empty
        // If they are then switch the default to visualizations
        viewBy = viewByHelpers.checkForEmptyArticlesOrInsights(viewBy, storiesArray.length === 0, teamDescription.pages.length === 0);

        var filterObj = func.filterObjFromQueryParams(req.query);

        if (process.env.NODE_ENV === 'enterprise' && process.env.SUBTEAMS) {
            baseUrl = baseUrl.replace(process.env.SUBDOMAIN, teamDescription.subdomain);
        }

        sourceDescriptions = sourceDescriptions.filter(function (description) {
            return description !== null;
        });

        each(teamDescription.pages, function (page) {
            page.baseUrl = baseUrl;
        });

        const sortedStories = orderBy(storiesArray, ['createdAt'], ['desc']);

        const pageNumber = req.query.page ? parseInt(req.query.page, 10) : 1;
        // Paginate datasets
        const paginatedVisualizations = paginateData(sourceDescriptions, viewBy.visualizations, VISUALIZATION_LIMIT, pageNumber);
        const paginatedInsights = paginateData(sortedStories, viewBy.insights, INSIGHTS_LIMIT, pageNumber);

        var data = {
            env: process.env,
            visualizations: paginatedVisualizations,
            team,
            baseUrl: baseUrl,
            routePath_base: baseUrl,
            filterObj: filterObj,
            displayTitleOverrides: { vizName: 'Visualization', teamName: 'Team' },
            insights: paginatedInsights,
            viewBy: viewBy || 'visualization',
            articles: teamDescription.pages,
            totalVizPages: Math.ceil(sourceDescriptions.length / VISUALIZATION_LIMIT),
            totalInsightsPages: Math.ceil(storiesArray.length / INSIGHTS_LIMIT),
            currentVizPage: viewBy.visualizations ? pageNumber : 1,
            currentInsightsPage: viewBy.insights ? pageNumber : 1,
            has_google_analytics: !!process.env.GOOGLE_ANALYTICS_ID,
            globalLibraries: [
                ...libraries.JAVASCRIPT,
            ],
        };

        if (req.user) {
            User.findById(req.user)
                .populate('defaultLoginTeam')
                .exec(function (err, user) {
                    if (err) {
                        return callback(err);
                    }
                    data.user = user;
                    callback(err, data);
                });
        } else {
            callback(err, data);
        }
    };

    const datasourceDescriptions = teamDescription.datasourceDescriptions
        .map((dataset) => {
            dataset.createdAt = get(dataset, 'previous_datasets[0].createdAt') || dataset.createdAt;
            return dataset;
        })
        .sort((a, b) => compareCreatedAt(b, a));

    async.map(datasourceDescriptions, iterateeFn, completionFn);
};

function paginateData(data, current, limit, page) {
    const pageNumber = current ? page : 1;
    const skipNResults = limit * (Math.max(pageNumber, 1) - 1);
    if (Array.isArray(data)) {
        const paginatedData = data.slice(skipNResults, skipNResults + limit);

        return paginatedData;
    }

    return data;
}
