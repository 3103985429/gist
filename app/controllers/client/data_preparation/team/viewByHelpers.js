/**
 * Instantiates the ViewBy based on the requested path
 */
module.exports.instantiateViewBy = (path) => {
    return {
        insights: path === '/insights' ? 'active' : '',
        articles: path === '/articles' ? 'active' : '',
        visualizations: path === '/' || path === '/visualizations' ? 'active' : ''
    };
};

/**
 * Checks to see if there any insights / articles
 * If there are none, then it switches active to visualizations
 * @param {Object} viewBy
 */
module.exports.checkForEmptyArticlesOrInsights = (viewBy, storiesNotActive, articlesNotActive) => {
    if (storiesNotActive && viewBy.insights === 'active') {
        viewBy.insights = '';
        viewBy.visualizations = 'active';
    } else if (articlesNotActive && viewBy.articles === 'active') {
        viewBy.articles = '';
        viewBy.visualizations = 'active';
    }
    return viewBy;
};

/**
 * Based on the viewBy, modify the baseurl to include the tab
 */
module.exports.modifyBaseUrl = (viewBy, baseUrl) => {
    if (viewBy.insights === 'active') {
        baseUrl += '/insights';
    } else if (viewBy.articles === 'active') {
        baseUrl += '/articles';
    } else {
        baseUrl += '/visualizations';
    }

    return baseUrl;
};
