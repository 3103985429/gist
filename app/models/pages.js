const mongoose_client = require('./mongoose_client');
const { isNull } = require('lodash');

const mongoose = mongoose_client.mongoose;
const Schema = mongoose.Schema;

// TODO:
// the acknowledgement should be moved into an object that's connected with the image

const pageSchema = Schema({
    pageTitle: String,
    slug: String,
    metaTitle: String,
    metaDescription: String,
    rawSummary: String,
    parsedSummary: String,
    rawBody: String,
    parsedBody: String,
    rawAbout: String,
    parsedAbout: String,
    url: String,
    thumbnail: String,
    image: String,
    pdf: String,
    acknowledgement: String,
    published: { type: Boolean, default: false },
    displayOnTeamPage: { type: Boolean, default: false },
    displayOnExplorePage: { type: Boolean, default: false },
    updatedAt: { type: Date, default: Date.now },
    createdAt: { type: Date, default: Date.now },
    isArticle: Boolean,
    state: String,
    team: { type: Schema.Types.ObjectId, ref: 'Team' },
    website: { type: Schema.Types.ObjectId, ref: 'Website' },
    createdBy: { type: Schema.Types.ObjectId, ref: 'User' },
    updatedBy: { type: Schema.Types.ObjectId, ref: 'User' },
}, { timestamps: { updatedAt: 'updatedAt' } });

const deepPopulate = require('mongoose-deep-populate')(mongoose);

pageSchema.plugin(deepPopulate);

pageSchema.pre('save', function (next) {
    this.updatedAt = Date.now();
    next();
});

pageSchema.statics.buildTeamPageQuery = (user) => {
    const baseQuery = { displayOnTeamPage: true };

    if (!user) {
        return {
            ...baseQuery,
            published: true,
        };
    }

    if (user.isSuperAdmin()) {
        return baseQuery;
    }

    const userId = user._id.toString();

    return {
        ...baseQuery,
        $or: [
            {
                team: {
                    $in: user._team
                        .filter(team => team.admin.some(id => id.toString() === userId))
                        .map(team => team._id),
                },
            },
            { _id: { $in: user._articleEditors } },
            { _id: { $in: user._articleViewers } },
            { published: true },
        ],
    };
};

pageSchema.statics.getShowcasedEntries = async function () {
    const pages = await this
        .find({
            published: true,
            state: 'approved',
            displayOnExplorePage: true,
        })
        .sort('-createdAt')
        .populate([
            {
                path: 'createdBy',
                select: 'firstName lastName -_id',
            },
            {
                path: 'website',
                select: 'slug -_id',
            },
            {
                path: 'team',
                select: 'subdomain -_id',
            },
        ]);

    return pages.filter(page => !isNull(page.team));
};

const Page = mongoose.model('Page', pageSchema);

module.exports = Page;
