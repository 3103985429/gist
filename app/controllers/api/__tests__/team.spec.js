import request from 'supertest';
import app from '../../../../app';
import { hydrateDb, getId, loginAndGetSessionCookie } from '../../../../internals/testing/backend/utils';
import {
    user1,
    user2,
    editorUser,
    viewerUser,
    superUser,
    hackyUser,
    articleEditorUser,
    articleViewerUser,
    jsonEditorUser,
} from '../../../../internals/testing/backend/fixtures/users';
import { team1 } from '../../../../internals/testing/backend/fixtures/teams';
import {
    simpsons,
    teamViz,
    privateVizWithEditor,
    apiDataset,
} from '../../../../internals/testing/backend/fixtures/datasets';
import { homepage, page1, article1 } from '../../../../internals/testing/backend/fixtures/pages';
import Team from '../../../models/teams';

import { getTeamBySubdomain } from '../team';


const promisified_getTeamBySubdomain = (req, callback) => new Promise(resolve => {
    getTeamBySubdomain(req, (...args) => {
        callback(...args);
        resolve();
    });
});

describe('Team controller', () => {
    let appRequest;

    beforeEach(async () => {
        await hydrateDb();
        appRequest = request(app);
    });

    describe('getTeamBySubdomain', () => {
        it('should return all listed visualizations for superadmin', async () => {
            const callback = jest.fn();
            const req = {
                user: getId(superUser),
                subdomains: ['user'],
                query: {},
            };

            await promisified_getTeamBySubdomain(req, callback);

            expect(callback.mock.calls[0][1][0].datasourceDescriptions.map(description => description._id)).toEqual([
                apiDataset._id,
                privateVizWithEditor._id,
                teamViz._id,
                simpsons._id,
            ]);
        });

        it('should return only public visualizations for non-team member', async () => {
            const callback = jest.fn();
            const req = {
                user: getId(hackyUser),
                subdomains: ['user'],
                query: {},
            };

            await promisified_getTeamBySubdomain(req, callback);

            expect(callback.mock.calls[0][1][0].datasourceDescriptions.map(description => description._id)).toEqual([
                apiDataset._id,
                simpsons._id,
            ]);
        });

        it('should return only public visualizations for unauthorized user', async () => {
            const callback = jest.fn();
            const req = {
                user: null,
                subdomains: ['user'],
                query: {},
            };

            await promisified_getTeamBySubdomain(req, callback);

            expect(callback.mock.calls[0][1][0].datasourceDescriptions.map(description => description._id)).toEqual([
                apiDataset._id,
                simpsons._id,
            ]);
        });

        it('should return all listed visualizations for team admin', async () => {
            const callback = jest.fn();
            const req = {
                user: getId(user2),
                subdomains: ['user'],
                query: {},
            };

            await promisified_getTeamBySubdomain(req, callback);

            expect(callback.mock.calls[0][1][0].datasourceDescriptions.map(description => description._id)).toEqual([
                apiDataset._id,
                privateVizWithEditor._id,
                teamViz._id,
                simpsons._id,
            ]);
        });

        it('should return visualizations that user is editor of and public ones for non-admin team member', async () => {
            const callback = jest.fn();
            const req = {
                user: getId(editorUser),
                subdomains: ['user'],
                query: {},
            };

            await promisified_getTeamBySubdomain(req, callback);

            expect(callback.mock.calls[0][1][0].datasourceDescriptions.map(description => description._id)).toEqual([
                apiDataset._id,
                privateVizWithEditor._id,
                simpsons._id,
            ]);
        });

        it('should return visualizations that user is viewer of and public ones for non-admin team member', async () => {
            const callback = jest.fn();
            const req = {
                user: getId(viewerUser),
                subdomains: ['user'],
                query: {},
            };

            await promisified_getTeamBySubdomain(req, callback);

            expect(callback.mock.calls[0][1][0].datasourceDescriptions.map(description => description._id)).toEqual([
                apiDataset._id,
                privateVizWithEditor._id,
                simpsons._id,
            ]);
        });

        it('should return pages with "displayOnTeamPage" flag for team admin', async () => {
            const callback = jest.fn();
            const req = {
                user: getId(user1),
                subdomains: ['glitter'],
                query: {},
            };

            await promisified_getTeamBySubdomain(req, callback);

            expect(callback.mock.calls[0][1][0].pages.map(page => page._id)).toEqual([
                homepage._id,
                article1._id,
                page1._id,
            ]);
        });

        it('should return pages with "displayOnTeamPage" flag for super admin', async () => {
            const callback = jest.fn();
            const req = {
                user: getId(superUser),
                subdomains: ['glitter'],
                query: {},
            };

            await promisified_getTeamBySubdomain(req, callback);

            expect(callback.mock.calls[0][1][0].pages.map(page => page._id)).toEqual([
                homepage._id,
                article1._id,
                page1._id,
            ]);
        });

        it('should return public ("published" flag) pages for non-team member', async () => {
            const callback = jest.fn();
            const req = {
                user: getId(hackyUser),
                subdomains: ['glitter'],
                query: {},
            };

            await promisified_getTeamBySubdomain(req, callback);

            expect(callback.mock.calls[0][1][0].pages.map(page => page._id)).toEqual([
                homepage._id,
            ]);
        });

        it('should return public ("published" flag) pages for unauthorized user', async () => {
            const callback = jest.fn();
            const req = {
                user: null,
                subdomains: ['glitter'],
                query: {},
            };

            await promisified_getTeamBySubdomain(req, callback);

            expect(callback.mock.calls[0][1][0].pages.map(page => page._id)).toEqual([
                homepage._id,
            ]);
        });

        it('should return pages that user is editor of and public ones for non-admin team member', async () => {
            const callback = jest.fn();
            const req = {
                user: getId(articleEditorUser),
                subdomains: ['glitter'],
                query: {},
            };

            await promisified_getTeamBySubdomain(req, callback);

            expect(callback.mock.calls[0][1][0].pages.map(page => page._id)).toEqual([
                homepage._id,
                article1._id,
            ]);
        });

        it('should return pages that user is viewer of and public ones for non-admin team member', async () => {
            const callback = jest.fn();
            const req = {
                user: getId(articleViewerUser),
                subdomains: ['glitter'],
                query: {},
            };

            await promisified_getTeamBySubdomain(req, callback);

            expect(callback.mock.calls[0][1][0].pages.map(page => page._id)).toEqual([
                homepage._id,
                article1._id,
            ]);
        });
    });

    describe('addAdmin', () => {
        it('should not allow to perform action when not an admin', async () => {
            const sessionCookie = await loginAndGetSessionCookie(appRequest, jsonEditorUser);
            const response = await appRequest
                .put(`/api/team/admin/${jsonEditorUser._id}`)
                .set('Cookie', sessionCookie);

            expect(response).toMatchSnapshot();
        });

        it('should not allow to perform action when user doesn\'t belong to the team', async () => {
            const sessionCookie = await loginAndGetSessionCookie(appRequest, user1);
            const response = await appRequest
                .put(`/api/team/admin/${user2._id}`)
                .set('Cookie', sessionCookie);

            expect(response).toMatchSnapshot();
        });

        it('should not allow to perform action when user is already an admin', async () => {
            const sessionCookie = await loginAndGetSessionCookie(appRequest, user1);
            const response = await appRequest
                .put(`/api/team/admin/${user1._id}`)
                .set('Cookie', sessionCookie);

            expect(response).toMatchSnapshot();
        });

        it('should add admin', async () => {
            const sessionCookie = await loginAndGetSessionCookie(appRequest, user1);
            const response = await appRequest
                .put(`/api/team/admin/${jsonEditorUser._id}`)
                .set('Cookie', sessionCookie);

            expect(response).toMatchSnapshot();

            const team = await Team.findById(team1._id);
            expect(team.admin).toEqual(expect.arrayContaining([jsonEditorUser._id]));
        });
    });

    describe('deleteAdmin', () => {
        it('should not allow to perform action when not an admin', async () => {
            const sessionCookie = await loginAndGetSessionCookie(appRequest, jsonEditorUser);
            const response = await appRequest
                .delete(`/api/team/admin/${jsonEditorUser._id}`)
                .set('Cookie', sessionCookie);

            expect(response).toMatchSnapshot();
        });

        it('should not allow to perform action when user doesn\'t belong to the team', async () => {
            const sessionCookie = await loginAndGetSessionCookie(appRequest, user1);
            const response = await appRequest
                .delete(`/api/team/admin/${user2._id}`)
                .set('Cookie', sessionCookie);

            expect(response).toMatchSnapshot();
        });

        it('should not allow to perform action when selected user is not an admin', async () => {
            const sessionCookie = await loginAndGetSessionCookie(appRequest, user1);
            const response = await appRequest
                .delete(`/api/team/admin/${jsonEditorUser._id}`)
                .set('Cookie', sessionCookie);

            expect(response).toMatchSnapshot();
        });

        it('should not allow to perform action when only one admin is in the team', async () => {
            const sessionCookie = await loginAndGetSessionCookie(appRequest, user1);
            const response = await appRequest
                .delete(`/api/team/admin/${user1._id}`)
                .set('Cookie', sessionCookie);

            expect(response).toMatchSnapshot();
        });

        it('should remove admin', async () => {
            await Team.findOneAndUpdate({ _id: team1._id }, { admin: [user1._id, jsonEditorUser._id] });

            const sessionCookie = await loginAndGetSessionCookie(appRequest, user1);
            const response = await appRequest
                .delete(`/api/team/admin/${jsonEditorUser._id}`)
                .set('Cookie', sessionCookie);

            expect(response).toMatchSnapshot();

            const team = await Team.findById(team1._id);
            expect(team.admin).not.toEqual(expect.arrayContaining([jsonEditorUser._id]));
        });
    });
});
