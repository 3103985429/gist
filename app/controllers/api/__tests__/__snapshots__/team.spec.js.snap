// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Team controller addAdmin should add admin 1`] = `
Object {
  "statusCode": 204,
  "text": "",
}
`;

exports[`Team controller addAdmin should not allow to perform action when not an admin 1`] = `
Object {
  "body": Object {
    "error": "Non-public error: Not an admin of the team",
  },
  "statusCode": 403,
}
`;

exports[`Team controller addAdmin should not allow to perform action when user doesn't belong to the team 1`] = `
Object {
  "body": Object {
    "error": "Non-public error: User doesn't belong to the team",
  },
  "statusCode": 400,
}
`;

exports[`Team controller addAdmin should not allow to perform action when user is already an admin 1`] = `
Object {
  "body": Object {
    "error": "Non-public error: User is already an admin",
  },
  "statusCode": 400,
}
`;

exports[`Team controller deleteAdmin should not allow to perform action when not an admin 1`] = `
Object {
  "body": Object {
    "error": "Non-public error: Not an admin of the team",
  },
  "statusCode": 403,
}
`;

exports[`Team controller deleteAdmin should not allow to perform action when only one admin is in the team 1`] = `
Object {
  "body": Object {
    "error": "Team needs to have at least one admin",
  },
  "statusCode": 400,
}
`;

exports[`Team controller deleteAdmin should not allow to perform action when selected user is not an admin 1`] = `
Object {
  "body": Object {
    "error": "Non-public error: User isn't an admin",
  },
  "statusCode": 400,
}
`;

exports[`Team controller deleteAdmin should not allow to perform action when user doesn't belong to the team 1`] = `
Object {
  "body": Object {
    "error": "Non-public error: User isn't an admin",
  },
  "statusCode": 400,
}
`;

exports[`Team controller deleteAdmin should remove admin 1`] = `
Object {
  "statusCode": 204,
  "text": "",
}
`;
